### DISCUSSION TITLE
Help producing the smallest Docker image footprint

### DISCUSSION THREAD

**Discussion by florianldt** [vercel/turborepo#5476] [Help] [answered]
I am pretty new to Turbo and I have a monorepo following the following structure:

```
▾ apps/
  ▸ api/
  ▸ web/
▸ node_modules/
▾ packages/
  ▸ eslint-config-custom/
  ▸ models/
  ▸ openapi/
  ▸ tsconfig/
.eslintrc.js
package-lock.json
package.json
README.md
turbo.json
```

I read the `with-docker` example but I find that this repo is not optimized for production and is not trying to minimize the image footprint (correct me if I am wrong).

So here is part of my Dockerfile for the `api` app:

```dockerfile
FROM node:16.16.0-alpine3.16 as base
RUN apk add --no-cache libc6-compat ncdu
RUN apk update
WORKDIR /app

FROM base AS builder
RUN npm i -g turbo
COPY . .
RUN turbo prune --scope=api --docker

FROM base AS installer
COPY --from=builder /app/out/json/ .
COPY --from=builder /app/out/package-lock.json ./package-lock.json
RUN npm ci --ignore-scripts=true
COPY --from=builder /app/out/full/ .
COPY turbo.json turbo.json
RUN npx -w api prisma generate
RUN npm run -w api build
RUN npm prune --omit=dev

FROM base AS runner
COPY --from=installer ./app/node_modules node_modules/
COPY --from=installer ./app/apps/api/dist apps/api/
COPY --from=installer ./app/apps/api/public apps/api/public/

EXPOSE 8000
CMD ["node", "apps/api/src/index.js"]
```

Everything is running as expected but the image size is still not optimum.

<img width="447" alt="Screenshot 2023-07-07 at 11 28 12 PM" src="https://github.com/vercel/turbo/assets/9365651/ecfaf326-451e-43ac-aa9c-5422073fc3df">

<img width="563" alt="Screenshot 2023-07-07 at 11 28 29 PM" src="https://github.com/vercel/turbo/assets/9365651/f185c790-b171-4ebc-8298-eeccf458b55a">

Even after removing the `packages` folder (they are not needed after the building phase), I should still be able to save at least 100MB from `typescript` + `@prisma`?

I did some tests outside docker and one thing I notice is that after running:

```
turbo prune --scope="api" --docker
cd out/json
npm i --omit=dev --ignore-script=true
```

I did notice duplicates of devDependencies (typescript in this case) increasing the size of the final image:

`out/`
<img width="558" alt="Screenshot 2023-07-07 at 11 07 50 PM" src="https://github.com/vercel/turbo/assets/9365651/e5042dfb-4469-49a2-998d-f8af91cefe57">

`out/json/`
<img width="501" alt="Screenshot 2023-07-07 at 11 08 27 PM" src="https://github.com/vercel/turbo/assets/9365651/aa8d6514-aa16-4f59-907f-030a524c11e9">

`out/json/node_modules/`
<img width="670" alt="Screenshot 2023-07-07 at 11 08 05 PM" src="https://github.com/vercel/turbo/assets/9365651/cb403be7-7d5a-4e82-ab59-f8b7b62bdf79">

`out/json/packages/eslint-config-custom/node_modules/`
<img width="501" alt="Screenshot 2023-07-07 at 11 08 27 PM" src="https://github.com/vercel/turbo/assets/9365651/551bebd2-ae06-4ca1-81e3-70407561e708">

`out/json/packages/models/node_modules/`
<img width="501" alt="Screenshot 2023-07-07 at 11 08 27 PM" src="https://github.com/vercel/turbo/assets/9365651/551bebd2-ae06-4ca1-81e3-70407561e708">

Dependency list of `api`:

```json
{
    "devDependencies": {
        "@types/cors": "^2.8.13",
        "@types/express": "^4.17.17",
        "@types/morgan": "^1.9.4",
        "@types/multer": "^1.4.7",
        "@types/multer-s3": "^3.0.0",
        "@types/node": "^20.2.5",
        "@types/uuid": "^9.0.2",
        "@typescript-eslint/eslint-plugin": "^5.59.8",
        "@typescript-eslint/parser": "^5.59.8",
        "dotenv-cli": "^7.2.1",
        "eslint": "^8.42.0",
        "eslint-config-airbnb-base": "^15.0.0",
        "eslint-config-airbnb-typescript": "^17.0.0",
        "eslint-config-prettier": "^8.8.0",
        "eslint-plugin-import": "^2.27.5",
        "eslint-plugin-prettier": "^4.2.1",
        "nodemon": "^2.0.22",
        "prisma": "^4.15.0",
        "rimraf": "^5.0.1",
        "ts-node": "^10.9.1",
        "typescript": "^5.1.3"
    },
    "dependencies": {
        "@aws-sdk/client-s3": "^3.360.0",
        "@prisma/client": "^4.15.0",
        "@resvg/resvg-js": "^2.4.1",
        "cors": "^2.8.5",
        "dotenv": "^16.1.4",
        "express": "^4.18.2",
        "express-async-errors": "^3.1.1",
        "firebase-admin": "^11.9.0",
        "models": "*",
        "morgan": "^1.10.0",
        "multer": "^1.4.5-lts.1",
        "multer-s3": "^3.0.1",
        "satori": "^0.10.1",
        "uuid": "^9.0.0",
        "winston": "^3.9.0",
        "zod": "^3.21.4"
    }
}
```

Dependency list of `models` (just types for shared dtos):

```json
{
    "devDependencies": {
        "eslint": "^7.32.0",
        "eslint-config-custom": "*",
        "tsconfig": "*",
        "typescript": "^4.5.2"
    }
}
```

Dependency list of `eslint-config-custom` (just types for shared dtos):

```json
{
  "dependencies": {
    "eslint": "^7.23.0",
    "eslint-config-next": "13.0.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-react": "7.31.8",
    "eslint-config-turbo": "latest"
  },
  "devDependencies": {
    "typescript": "^4.7.4"
  }
}
```

Anything I am missing or not understanding why typescript is listed in the root dependencies but also in the sub ones even if I am installing only the production dependencies? Hoping that it is clear enough to understand.

**Comment by anthonyshew** [comment, marked as answer]
Looks like you're doing good work here!

There most likely are some opportunities for `turbo prune` to create slimmer builds. I'm seeing most of the issues you're mentioning as mentioned in issues on the repo. If you'd like to get those issues more support, please add reactions to them so we know to prioritize them. 🙏
